mlflow/sklearn/__init__.py

```diff
@@ -794,24 +805,29 @@ def _log_posttraining_metadata(estimator, *args, **kwargs):
         # log common metrics and artifacts for estimators (classifier, regressor)
         _log_specialized_estimator_content(estimator, mlflow.active_run().info.run_id, args, kwargs)
 
-        input_example = None
-        signature = None
-        if hasattr(estimator, "predict"):
-            try:
-                # Fetch an input example using the first several rows of the array-like
-                # training data supplied to the training routine (e.g., `fit()`)
-                fit_arg_names = _get_arg_names(estimator.fit)
-                X_var_name, y_var_name = fit_arg_names[:2]
-                input_example = _get_Xy(args, kwargs, X_var_name, y_var_name)[0][
-                    :INPUT_EXAMPLE_SAMPLE_ROWS
-                ]
-
-                model_output = estimator.predict(input_example)
-                signature = infer_signature(input_example, model_output)
-            except Exception as e:  # pylint: disable=broad-except
-                input_example = None
-                msg = "Failed to infer an input example and model signature: " + str(e)
-                _logger.warning(msg)
+        def get_input_example():
+            # Fetch an input example using the first several rows of the array-like
+            # training data supplied to the training routine (e.g., `fit()`)
+            fit_arg_names = _get_arg_names(estimator.fit)
+            X_var_name, y_var_name = fit_arg_names[:2]
+            input_example = _get_Xy(args, kwargs, X_var_name, y_var_name)[0][
+                :INPUT_EXAMPLE_SAMPLE_ROWS
+            ]
+            return input_example
+
+        def infer_model_signature(input_example):
+            if not hasattr(estimator, "predict"):
+                raise Exception("the model used has no predict function")
+
+            return infer_signature(input_example, estimator.predict(input_example))
+
+        input_example, signature = resolve_input_example_and_signature(
+            get_input_example,
+            infer_model_signature,
+            log_input_example,
+            log_model_signature and hasattr(estimator, "predict"),
```

Comment:
We should remove this line, otherwise signature collection fails silently.

Reply:
oops sorry i thought i did this :(